### DISCUSSION TITLE
[Feature]: Add federation to share albums with users from other immich servers

### DISCUSSION THREAD

**Discussion by anbraten** [immich-app/immich#13621] [Feature Request]
### Feature detail

As a user I would like to share images. In addition to sharing with users of my instance, public users via link (see #265), I would like to be able to share with a user from another immich instance.

__What's the benefit over "simply" sharing with a link?__

- The other user could be allowed to edit the album
- By sharing with an explicit user from an other instance this user would be able to sync the shared photos to their instance as well

### Platform

Server

**Comment by bo0tzz** [comment]
Thanks for the suggestion! Federation between instances is not something that we currently consider in scope for Immich. As for the specific features you mention, editing shared albums as an external user is something that is already planned in that feature. Syncing photos from a shared album is you could accomplish with a fairly small external script using the Immich API.

**Comment by domacikolaci** [comment]
Any chance this could reenter the roadmap?

With FUTO funding and the project coming such a long way, it would be amazing to see this on the roadmap even as a "down the line" consideration.

I would love to have the ability to "connect" my server with my mates server so we can share albums and data between each other and allow our users to share data between each other. For anyone from a corporate Microsoft background, this B2B integrations with SharePoint and Teams as an example.

**Comment by axgd-code** [comment]
I am 100% on board with saying that it's missing from the current roadmap. Google's shared album features are super attractive to users. A similar feature in Immich is more complex because we are in a multi-instance system. One of the Immich members talks about it in this conference: https://www.youtube.com/watch?v=F_2oydlnpCQ, so I remain hopeful. It's the feature I also miss to use Immich on a daily basis.

**Comment by MrModest** [comment]
Just willing to share my vision of the implementation of this feature, if the team ever will have time for this :D

https://github.com/immich-app/immich/discussions/13620

**Comment by bo0tzz** [comment]
Reopened as @zackpollard has expressed a desire to work on this at some point.

**Comment by looowizz** [comment]
Might be interesting to try to implement this using the [Open Cloud Mesh Protocol ](https://github.com/cs3org/OCM-API) for compatibility with Nextcloud and OwnCloud?

**Reply by MrModest** [reply]
How do you see Immich communicating with a File Cloud app? 🤔

**Reply by looowizz** [reply]
Nextcloud currently has a similar photos backup solution called [Memories](https://memories.gallery/). This just uses a photos folder, ~then folders for albums within that~.

I'd imagine Immich could work in a similar way, sharing an album Could work function as sharing a folder. Might be a bit more complicated to implement as lots of the metadata for images are stored in Immich's database and sidecar files (which arent yet supported by Memories) but I think it'd be quite cool!
EDIT: Oops, seems like Memories doesn't just use folders for albums. I think, still, using this protocol would be good for the future, as it is probably one of the most well suited and widely used, standardised protocol for this.

**Reply by MrModest** [reply]
Immich (unlike Google) follows the approach "do not touch the original files", so as long as album share will cause sending a copy of original photos to the NC, all metadata should remain in the files directly. 
Immich stores metadata in the DB just for performance reasons, it doesn't remove them from files themselves.

**Reply by MrModest** [reply]
btw, checked their demo: https://demo.memories.gallery/apps/memories/

Not as bad, as I'd expected from a Nextcloud app :D all basic features are there and it even has an "Explore" feature like search and Face recognition. Immich is still a way better, in my opinion, thought 😁

**Reply by looowizz** [reply]
> Immich (unlike Google) follows the approach "do not touch the original files", so as long as album share will cause sending a copy of original photos to the NC, all metadata should remain in the files directly. Immich stores metadata in the DB just for performance reasons, it doesn't remove them from files themselves.

Ahh i see, yeah. I guess the only way would be to make an abstraction so that federating cloud services interacting with the photos (e.g. changing metadata, deleting files) just invokes the corresponding function in Immich. Not sure how possible this is, as not familiar with the OCM protocol. Plus, it will likely be a lot more work than just using a custom (hopefully open) API, so understandable why the devs wouldn't want to do this. :)

**Reply by isZYKerman** [reply]
Well, I use both nextcloud for file management and immich for photo management. Nextcloud can share files between servers in one click. For example, on my nextcloud server I generate a link for sharing a file and send the link to my friend. He can then open the link, fill in the blank with "his_username@his_server_address" and the files are successfully sent to his server directly. I believe that this is achieved though email notification because it just won't work if you have not set up the "send notification through email" function correctly. Considering that we also have similar email notification function in immich, maybe we can achieve synchronization between immich servers like this somehow...?

**Comment by max-tet** [comment]
I was just about to create a new discussion called _Cross-instance sharing/syncing_ when I found this one. This would be a very useful feature, but I have a bit of a different approach in mind. Why not leverage the existing sharing link feature?

The one who shares would create a sharing link just like it is possible right now. The receiver could take that link and embed its target into their own Immich instance. There could be a _embed_ or _mount_ button somewhere which prompts you to provide a sharing link to embed. The shared images would then be integrated with your own images. They could be labelled with a special icon so you see that they are not really yours. You could put them in Albums or organize them any way you like, just like if they were your own. Also, they could be either loaded on-demand from the foreign instance or synced one-way in the background or a mixture of both.

There would be no explicit _connecting_ of instances needed and no problem with instances that cannot be reached as long as sharing links work in your setup. Implementation would probably also be relatively straightforward: no work to implement the sharing side, only for the embed feature.

For the sake of transparency: I would love that feature because it would make it much more useful to use Immich on [Freeshard](https://freeshard.net/), of which I am the maintainer. Freeshard provides personal, single-user cloud VMs so every Immich instance is for a single user by definition. That means sharing between users is only possible through some kind of cross-instance sharing.

**Reply by Wateir** [reply]
> There would be no explicit connecting of instances needed and no problem with instances that cannot be reached as long as sharing links work in your setup. Implementation would probably also be relatively straightforward: no work to implement the sharing side, only for the embed feature.

Just a local copy ot the thumbmail on the local server to don't make a bunch of call to other instance when you are just browsing you're library

**Comment by eddified** [comment]
I found this discussion because I imagined this feature on my own separately and thought it would be cool. However, I do think we should prioritize the following over federation:

* re-doing partner sharing / same-instance sharing - Use case: allowing parents to manage all photos in the system
* enhancing external library integration
* improving AI tools: search and facial recognition

**Reply by bo0tzz** [reply]
@eddified please don't off-topic post in unrelated PRs about what features you think should be added instead.

**Comment by laxmanpradhan** [comment]
Anu chance this could be revisited now that stable is out. Seems like a great option as its common for selfhosters to know other self hosters who each would want their own servers.

**Reply by axgd-code** [reply]
it's in the roadmap :)

**Reply by bo0tzz** [reply]
Where?

**Reply by axgd-code** [reply]
https://youtu.be/OgLoixC597A?si=df74R85m9brGJf4X&t=1389

**Comment by smiba** [comment]
I want to voice that this has been the primary reason for not being able to let go of Google Photos for years now, immich does everything I need to do (and often better than competitors) but the lack of cross-instance collaboration makes it difficult for me and friends to fully switch over.

We usually make a shared album after a nights out or a trip and it's rather frictionless on GPhotos. Many of us have immich instances, but we just can't use it for easy collaboration

A public link doesn't work, as it doesn't leave us in control of our own photos. In GPhotos the user who owns the photo is able to remove it again, but not mess with the other photos

**Comment by tigattack** [comment]
I would love to see this. Similarly to others, I love Immich but still find myself attached to Google Photos for the sharing features.
This makes the whole experience super clunky since I need to upload to both and manually keep my albums in sync.
It would be fantastic to be able to share albums across servers and do away with all that "photo admin".

Since it was mentioned as a future feature almost a year ago (https://youtu.be/OgLoixC597A?t=1389), I hope to see it on the [roadmap](https://immich.app/roadmap) and/or tracked here on GitHub soon!

As an aside, this feels semi-related to #12614, perhaps as an extension or "v2" of that issue.

**Comment by lukeet332** [comment]
Just want to mirror a similar sentiment as above this project is amazing! but switching my family over fully simply won't be possible until some kind of shared album federation is enabled between multiple server instances 😢

**Reply by paucabrerah** [reply]
I believe this is a super overlooked feature. It's been requested for a while now and there's obvious real life benefits for a lot of users. 

Developers should really consider this into the next release @zackpollard

**Reply by lukeet332** [reply]
I've actually just made my own fully working add-on today that enables cross-server shared albums:
https://github.com/lukeet332/immich-shared-albums
Feel free to give it a try 😄

**Reply by paucabrerah** [reply]
It looks good but I would rather see it directly implemented within Immich. Why not using your code for a merge request into Immich own repo?

**Reply by lukeet332** [reply]
It's a sidecar my code is designed to work alongside Immich. I didn't really want to rush breaking my Immich instance or creating a fork to maintain that would never be merged lol

**Reply by paucabrerah** [reply]
That's great. But I still would prefer this to be directly implemented into Immich instead of patching with external containers. That's why I proposed you used that code to create a MR into Immich so that this "plugin" you made is integrated into it.
